src/CustomMarshalers/src/System/Runtime/InteropServices/CustomMarshalers/EnumVariantViewOfEnumerator.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Collections;
+using System.Runtime.InteropServices.ComTypes;
+
+namespace System.Runtime.InteropServices.CustomMarshalers
+{
+    [Serializable]
+    internal class EnumVariantViewOfEnumerator : IEnumVARIANT, ICustomAdapter, ICustomQueryInterface
+    {
+        private static readonly Guid IID_IManagedObject = new Guid("C3FCC19E-A970-11d2-8B5A-00A0C9B7C9C4");
+
+        public EnumVariantViewOfEnumerator(IEnumerator enumerator)
+        {
+            if (enumerator is null)
+            {
+                throw new ArgumentNullException(nameof(enumerator));
+            }
+
+            Enumerator = enumerator;
+        }
+        internal IEnumerator Enumerator { get; }
+
+        public IEnumVARIANT Clone()
+        {
+            if (Enumerator is ICloneable clonable)
+            {
+                return new EnumVariantViewOfEnumerator(clonable.Clone() as IEnumerator);
+            }
+            else
+            {
+                throw new COMException(SR.Arg_EnumNotCloneable, HResults.E_FAIL);
+            }
+        }
+
+        public int Next(int celt, object[] rgVar, IntPtr pceltFetched)
+        {
+            int numElements = 0;
+
+            try
+            {
+                if (celt > 0 && rgVar == null)
+                {
+                    return Marshal.GetHRForException(new ArgumentException());
```

Comment:
You don't need to do this, just throw the exception. The marshaller will convert well know Exceptions into correct `HRESULT` values. The other consideration here is `GetHRForException()` doesn't always do what you want it to do and in this case, bad things could occur if the user is already in an interop environment.

Reply:
Boo... returning `S_FALSE` means we can't throw.. 😞 You should add the needed `HRESULT` to the `HResults` type.